src/main/java/org/apache/sling/resourceresolver/impl/ResourceResolverFactoryActivator.java

```diff
@@ -128,6 +129,9 @@
 
     private volatile ResourceResolverFactoryConfig config = DEFAULT_CONFIG;
 
+    /** Alias path whitelist */
+    private AtomicReferenceArray<String> aliasPathAllowList;
```

Comment:
I think it's simpler if you have a `volatile` list here and for thread safety either make them immutable , e.g. `Collections.unmodifiableList(...)` or use a thread-safe list like `CopyOnWriteArrayList`.

Reply:
I strongly agree with @rombert. `AtomicReferenceArray` is not the correct data-structure for how you use it. It allows atomic updates to individual array entries, but not for the entire array. You need an atomic update of the entire array (or collection). Indeed, you reassign this field later, which is not thread-safe because the field is not volatile.

If you want to atomically swap out the entire configuration, you need a volatile field and make a local copy of the field before you work with it. Or you use a thread-safe data-structure, like e.g. `CopyOnWriteArrayList` as Robert suggests.

Reply:
Initially, Volite keyword is used but sonar bot suggestion suggested to use **AtomicReferenceArray** as should not have Non-primitive fields as "volatile" and throwing bug in the PR.

Reply:
I am not sure why a `volatile CopyOnWriteArrayList` wouldn't work:

- The `volatile` modifier ensures that that the readers of the field always get the latest value - so changing the reference works
- the `CopyOnWriteArrayList` implementation is thread-safe internally, so changes made by one thread are visible to other threads

Maybe SonarCloud flagged a volatile field holding a non-thread-safe list implementation, like `ArrayList`? That would not solve the second problem listed above.

I think using a List is much more easier to understand than an AtomicReferenceArray, so I'd prefer that if at all possible.

Reply:
I tried to list and volatile, in both cases, sonar cloud was throwing a bug saying "should not have Non-primitive fields as volatile, and suggested to use AtomicReferenceArray.
So I modified the PR and used AomicReferenceArray.

Reply:
I'd suggest to drop the volatile keyword, just use CopyOnWriteArrayList and initialize it here. Let the rest of the code never set it to another value. Just use add/remove/clear. Thread-safe, readable and no Sonar issues.

Reply:
The sonar message is misleading. A more accurate message would say that a volatile field should only hold an immutable object (which includes primitives). So wrapping a `List` using e.g. `Collections.unmodifiableList()` would also be correct and _should not_ trigger any sonar warnings. Alternatively, using a final field holding a thread-safe collection like `CopyOnWriteArrayList` is also correct (in the sense of thread-safety).

Swapping out an immutable object with another in a volatile field constitutes an atomic update, which is often desirable.

Updating a thread-safe collection is not necessarily atomic, so there _could_ be a time-window when the field could be read and the collections contents are in the middle of an update. This may or may not be a problem depending on the use-case.

Using `Atomic*Array` sounds like bad advice to me, because it usually (my opinion) you will not want to swap out the contents of a particular array index atomically. It allows elements to be updated atomically, which is not what you want here.

You want to swap out one configuration for another, so updating a volatile field with an immutable collection sounds most correct to me.